src/server/dep_graph.rs

```diff
@@ -0,0 +1,228 @@
+//! Dependency graph for tracking reverse include relationships between files.
+//!
+//! This module tracks which files include a given file, enabling "find references"
+//! to search dependent files. Forward dependencies (what a file includes) are
+//! already tracked in `ParsedCode.includes`.
+
+use std::collections::{HashMap, HashSet};
+
+use lsp_types::Url;
+
+/// Tracks which files include a given file (reverse dependency edges).
+///
+/// For file A that includes file B, `included_by[B]` contains A.
+/// This enables finding all files that might reference symbols from B.
+#[derive(Default)]
+pub struct DependencyGraph {
+    /// Reverse edges: file -> files that include it
+    included_by: HashMap<Url, HashSet<Url>>,
+    /// Track what each file includes, so we can update reverse edges on change
+    includes_cache: HashMap<Url, HashSet<Url>>,
+}
+
+impl DependencyGraph {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Update the graph when a file is parsed/changed.
+    /// Removes old reverse edges for this file and adds new ones.
+    pub fn update_file(&mut self, file: &Url, new_includes: &[Url]) {
+        // Remove old reverse edges
+        if let Some(old_includes) = self.includes_cache.remove(file) {
+            for old_include in old_includes {
+                if let Some(dependents) = self.included_by.get_mut(&old_include) {
+                    dependents.remove(file);
+                    if dependents.is_empty() {
+                        self.included_by.remove(&old_include);
+                    }
+                }
+            }
+        }
+
+        // Add new reverse edges
+        if !new_includes.is_empty() {
+            let includes_set: HashSet<Url> = new_includes.iter().cloned().collect();
+
+            for include in &includes_set {
+                self.included_by
+                    .entry(include.clone())
+                    .or_default()
+                    .insert(file.clone());
+            }
+
+            self.includes_cache.insert(file.clone(), includes_set);
+        }
+    }
+
+    /// Remove a file from the graph (when closed/deleted).
+    pub fn remove_file(&mut self, file: &Url) {
+        // Remove reverse edges where this file is the dependent
+        if let Some(old_includes) = self.includes_cache.remove(file) {
+            for old_include in old_includes {
```

Comment:
A bit more of a functional approach here might reduce the cyclomatic complexity here:


```suggestion
        for old_include in self.includes_cache.remove(file).iter().flatten() {
```